## Problem 10: Summation of primes

The sum of the primes below 10 is $2 + 3 + 5 + 7 = 17$.

Find the sum of all the primes below two million.

In [1]:
import numpy as np
import tqdm
import time

In [2]:
''' 
Determines whether the input number is a prime. This is done by seeing if the number is 
divisible by any primes. Only need to test primes that are smaller than the number's square root.
This requires the prime_list to include all primes up to the square root of the test number.

eg:
>>> is_prime_with_prime_list(7, [2,3,5])
>>> True
'''

def is_prime_with_prime_list(number : int, prime_list : list) -> bool:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        This function determines whether the input number is a prime. This is done by seeing if the number is 
        divisible by any primes. Only need to test primes that are smaller than the number's square root.
        This requires the prime_list to include all primes up to the square root of the test number.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : number : Number to check for primality.
        : list : prime_list : A list of prime numbers to use to check for divisibility.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : bool : The number is prime.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> is_prime_with_prime_list(7, [2,3,5])
        >>> True
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(number, int): raise ValueError('Please enter an integer > 0 for the number argument.')
    if number < 1: raise ValueError('Please enter an integer > 0 for the number argument.')
    if not isinstance(prime_list, list): raise ValueError('Please enter a list of integers for the prime_list argument.')

    for prime in prime_list:
        
        # Only need to test up to the square root of the number.
        if prime > np.sqrt(number) + 1:
            break
        
        # If the number is divisible by a prime, then it is not prime.
        if number % prime == 0:
            return False
    
    # If the number is not divible by a prime, it must be the next prime.
    return True

In [3]:
''' 
Produces a list of primes up to the max specified. It does this by testing all the numbers up to the max for primality.

eg:
>>> produce_prime_list_with_individual_testing(10)
>>> [2, 3, 5, 7]
'''

def produce_prime_list_with_individual_testing(max_number : int) -> list:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Produces a list of primes up to the max specified. It does this by testing all the numbers up to the max for primality.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : number : Number to check for primality.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : prime_list : List of prime numbers that are smaller than max_number.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> produce_prime_list_with_individual_testing(10)
        >>> [2, 3, 5, 7]
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(max_number, int): raise ValueError('Please enter an integer > 0 for the max_number argument.')
    if max_number < 1: raise ValueError('Please enter an integer > 0 for the max_number argument.')
    
    prime_list = []
    
    # Loop through all integers up to max
    for number in range(2, max_number):
        
        # Test for primality
        if is_prime_with_prime_list(number, prime_list): prime_list.append(number)
    
    return prime_list

In [4]:
max_number = 2_000_000

start_time = time.time()

# Produce list of primes.
primes = produce_prime_list_with_individual_testing(max_number)

# Sum primes
solution = 0
for prime in primes:
    solution += prime
end_time = time.time()
    
print(f'There are {len(primes):,} primes below {max_number:,} and their sum is {solution:,}. This was calculated in {end_time - start_time:.5f}s')

There are 148,933 primes below 2,000,000 and their sum is 142,913,828,922. This was calculated in 35.23885s


The above method for compiling a list of primes is quite slow. A quicker method is The Sieve of Eratosthenes:

The sieve of Eratosthenes is an ancient algorithm for finding all prime numbers up to any given limit. It does so by iteratively marking as composite (i.e., not prime) the multiples of each prime, starting with the first prime number, 2. The multiples of a given prime are generated as a sequence of numbers starting from that prime, with constant difference between them that is equal to that prime. This is the sieve's key distinction from using trial division to sequentially test each candidate number for divisibility by each prime. Once all the multiples of each discovered prime have been marked as composites, the remaining unmarked numbers are primes.

In [5]:
''' 
This function sieves an array of primality flags using the number provided.
It sets the primality flag to 0 for all numbers that are multiples of the number.

eg:
>>> sieve(2, [1, 1, 1, 1, 1, 1, 1, 1, 1])
>>> [1, 1, 0, 1, 0, 1, 0, 1, 0]
'''

def sieve(number : int, flags : list) -> list:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        This function sieves an array of primality flags using the number provided.
        It sets the primality flag to 0 for all numbers that are multiples of the number.
        
        The sieve of Eratosthenes is an ancient algorithm for finding all prime numbers up to any given limit. 
        It does so by iteratively marking as composite (i.e., not prime) the multiples of each prime, starting 
        with the first prime number, 2. The multiples of a given prime are generated as a sequence of numbers 
        starting from that prime, with constant difference between them that is equal to that prime. This is the 
        sieve's key distinction from using trial division to sequentially test each candidate number for divisibility
        by each prime. Once all the multiples of each discovered prime have been marked as composites, the remaining 
        unmarked numbers are primes.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : number : Number to set flags for multiples.
        : list : flags : The list of primality flags.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : flags : The list of primality flags.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> sieve(2, [1, 1, 1, 1, 1, 1, 1, 1, 1])
        >>> [1, 1, 0, 1, 0, 1, 0, 1, 0]
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(number, int): raise ValueError('Please enter an integer > 0 for the number argument.')
    if number < 1: raise ValueError('Please enter an integer > 0 for the number argument.')
    

    # Keep setting the flags to 0 until we reach the end of the list.
    # Only need to multiply our number until it reaches the end of the list.
    for multiple in range(2, int(np.ceil(len(flags) / number) + 1)):
        
        # Don't go beyond the length of the list.
        if (number * multiple) - 2 < len(flags):
            
            # Set the primalilty flag to 0.
            # Need to convert number to flag index:
            # 0 index is the number 2
            # 1 index is the number 3
            # Conversion: index = number -2
            flags[(number*multiple) - 2] = 0
    
    return flags

In [6]:
''' 
Produces an array of prime flags which store the primality of the integers 2 to max_number - 1.
The sieve is run and the prime flags are set. Then a list of numbers is produced containing only numbers with prime flags.

eg:
>>> produce_primality_with_sieve(10)
>>> array([1., 1., 0., 1., 0., 1., 0., 0.])
'''

def produce_primality_with_sieve(max_number : int) -> list:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        This function produces an array of prime flags which store the primality of the integers 2 to max_number - 1.
        The sieve is run and the prime flags are set. Then a list of numbers is produced containing only numbers 
        with prime flags.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : max_number : Maximum number to include in prime sieve.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : prime_flags : The list of primality flags.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> produce_primality_with_sieve(10)
        >>> array([1., 1., 0., 1., 0., 1., 0., 0.])
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(max_number, int): raise ValueError('Please enter an integer > 0 for the max_number argument.')
    if max_number < 1: raise ValueError('Please enter an integer > 0 for the max_number argument.')
    
    # Flags are for numbers 2 to max_number-1
    prime_flags = np.ones(max_number-2)
    
    # Loop through the prime flags
    # Run the sieve when a prime is found
    for index in range(len(prime_flags)-1):
        
        # If the prime flag is still set, then it is a prime. Run the sieve
        if prime_flags[index]:
            prime_flags = sieve(index + 2, prime_flags)
    
    return prime_flags

In [7]:
''' 
Produces a list of primes using an array of primalilty flags.

eg:
>>> produce_prime_list_with_flags([1., 1., 0., 1., 0., 1., 0., 0.])
>>> [2, 3, 5, 7]
'''

def produce_prime_list_with_flags(flags : list) -> list:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Produces a list of primes using an array of primalilty flags.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : flags : The list of primality flags.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : primes : The list of primes.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> produce_prime_list_with_flags([1., 1., 0., 1., 0., 1., 0., 0.])
        >>> [2, 3, 5, 7]
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    primes = []
    
    for index in range(len(flags)-1):
        
        # If flag indicates prime, add prime to list.
        if flags[index]: primes.append(index + 2)
    
    return primes

In [8]:
''' 
Produces a list of primes using the sieve to produce an array of primality.

eg:
>>> produce_prime_list_with_sieve(20)
>>> [2, 3, 5, 7, 11, 13, 17]
'''

def produce_prime_list_with_sieve(max_number : int) -> list:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Produces a list of primes using the sieve to produce an array of primality.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : max_number : Maximum number to search for primes.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : prime_list : The list of primes.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> produce_prime_list_with_sieve(20)
        >>> [2, 3, 5, 7, 11, 13, 17]
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(max_number, int): raise ValueError('Please enter an integer > 0 for the max_number argument.')
    if max_number < 1: raise ValueError('Please enter an integer > 0 for the max_number argument.')
    
    flags = produce_primality_with_sieve(max_number)
    
    prime_list = produce_prime_list_with_flags(flags)
    
    return prime_list

In [9]:
max_number = 2_000_000

start_time = time.time()
primes = produce_prime_list_with_sieve(max_number)

solution = 0
for prime in primes:
    solution += prime
end_time = time.time()
    
print(f'There are {len(primes):,} primes below {max_number:,} and their sum is {solution:,}. This was calculated in {end_time - start_time:.5f}s')

There are 148,933 primes below 2,000,000 and their sum is 142,913,828,922. This was calculated in 1.98403s


### Problem 10 Solution:

There are 148,933 primes below 2,000,000 and their sum is 142,913,828,922.